### DISCUSSION TITLE
using classes of dependency plugin transitive library : java.lang.LinkageError:

### DISCUSSION THREAD

**Discussion by moglideveloper** [JetBrains/sbt-idea-plugin#107] [Q&A] [answered]
There is a **sample intellij plugin** that depends on **org.intellij.scala plugin** also.   
Now, this sample intellij plugin has a functionality to fetch json over http and then parse it.   

For parsing json in scala, there are multiple libraries, and we decided to use spray json library.   

---

**Reason for choosing spray json :-**

spray json library is present in library dependency graph of org.intellij.scala, **and sample intellij plugin directly depends on org.intellij.scala plugin**.   

So to keep **sample intellij plugin** library size minimum, we thought of using json parser library that is used by **org.intellij.scala plugin**.   

But, we are getting PluginClassLoader exception :-   

```
java.lang.LinkageError: loader constraint violation: when resolving method 'spray.json.RootJsonFormat spray.json.DefaultJsonProtocol$.jsonFormat1(scala.Function1, spray.json.JsonFormat, scala.reflect.ClassTag)' the class loader com.intellij.ide.plugins.cl.PluginClassLoader @2a4b0539 of the current class, com/sample/plugin/FetchDomainObjectsFromWire, and the class loader com.intellij.ide.plugins.cl.PluginClassLoader @573b8e9c for the method's defining class, spray/json/DefaultJsonProtocol$, have different Class objects for the type scala/Function1 used in the signature (com.sample.plugin.FetchDomainObjectsFromWire is in unnamed module of loader com.intellij.ide.plugins.cl.PluginClassLoader @2a4b0539, parent loader 'bootstrap'; spray.json.DefaultJsonProtocol$ is in unnamed module of loader com.intellij.ide.plugins.cl.PluginClassLoader @573b8e9c, parent loader 'bootstrap')

```	
---

**Our understanding so far for this exception :-**   

As org.intellij.scala and sample intellij are loaded in different classloaders, so sample intellij plugin can't use any class loaded by PluginClassLoader of org.intellij.scala.    
To use spray json classes in sample intellij plugin, spray json should be explicitly added to libraryDependencies in sample intellij plugin.   

Adding spray json to libraryDependencies in sample intellij plugin will solve the class-loader issue,     

but increase the packaged plugin binary size, as spray json library is also packaged in sample intellij plugin.   

**plugin.xml**

```
    <id>com.logicovercode</id>
    <name>IjPlugin</name>
    <vendor email="moglideveloper@gmail.com" url="https://github.com/JetBrains/sbt-idea-plugin/discussions/107">Linkage Error</vendor>
    <version>2021.1.18</version>
    <depends>org.intellij.scala</depends>
    <idea-version since-build="211.5538" until-build="211.*"/>
```

**build.sbt :**

```
scalaVersion := "2.13.2"

intellijPlugins += "org.intellij.scala:2021.1.18".toPlugin
intellijDownloadSources := true
ThisBuild / bundleScalaLibrary  := false

```

**plugins.sbt**

`addSbtPlugin("org.jetbrains" % "sbt-idea-plugin" % "3.12.2")`

---

Summary: if spray-json is not present in SamplePlugin/target/plugin/logicIj/lib, then plugin is giving LinkageError, when run from intellij-run configuration or by loading sample plugin zip in fresh intellij instance.

_Kindly confirm, if our understanding is correct and please suggest recommended approach/ best practices for this scenario._   

Thanks.

**Comment by unkarjedy** [comment]
I've created the dummy plugin via [IntelliJ Platform plugin template](https://blog.jetbrains.com/scala/2021/04/07/intellij-scala-plugin-2021-1/#intellij-platform-plugin-template)
I can create Spray entities and can see no LinkageErrors:
```scala
println(new spray.json.JsString("{}"))
```

Here is an example classloader for a simple action.
Note, that it contains two parent classloaders, including the classloader of the Scala Plugin.

![image](https://user-images.githubusercontent.com/3989292/117786762-86946f00-b24e-11eb-8a3f-88c03a6fd7d3.png)

>As org.intellij.scala and sample intellij are loaded in different classloaders, so sample intellij plugin can't use any class loaded by PluginClassLoader of org.intellij.scala

If you do not create any clasloaders explicitly and if your plugin's code is run in IDEA process you should be able to use the classes from Scala plugin, and the classes it depends on.
Unless those classes are used outside IDEA main process (thus use a separate classloader), which is not the case for the spray.

Here is full `build.sbt` for the dummy plugin I used:
```scala
import org.jetbrains.sbtidea.Keys._

lazy val testIntellij2 =
  project.in(file("."))
    .enablePlugins(SbtIdeaPlugin)
    .settings(
      version := "0.0.1-SNAPSHOT",
      scalaVersion := "2.13.2",
      ThisBuild / intellijPluginName := "TestIntellij2",
      ThisBuild / intellijBuild := "211.7142.45",
      ThisBuild / intellijPlatform := IntelliJPlatform.IdeaCommunity,
      Global / intellijAttachSources := true,
      Compile / javacOptions ++= "--release" :: "11" :: Nil,
      intellijPlugins += "com.intellij.properties".toPlugin,
      unmanagedResourceDirectories in Compile += baseDirectory.value / "resources",
      unmanagedResourceDirectories in Test += baseDirectory.value / "testResources",

      intellijPlugins += "org.intellij.scala:2021.1.18".toPlugin,
      intellijDownloadSources := true,
      ThisBuild / bundleScalaLibrary := false,
    )
```
and `plugin.xml`
```xml
<idea-plugin>
    <id>naumneko.test.vendor.testintellij2</id>
    <name>TestIntellij2</name>
    <version>0.0.1-SNAPSHOT</version>
    <vendor>naumneko.test.vendor</vendor>

    <description>A new plugin for IntelliJ platform written in Scala and built with SBT</description>

    <idea-version since-build="211.0"/>

    <depends>com.intellij.modules.platform</depends>
    <depends>com.intellij.properties</depends>
    <depends>org.intellij.scala</depends>
    
<!--    SOME ACTIONS, LISTENERS, etc... -->
</idea-plugin>

```

**Reply by moglideveloper** [reply]
Thanks @unkarjedy for the sample project.

**Comment by unkarjedy** [comment]
@moglideveloper 
Do you create any classloaders explicitly?
At which point in the plugin are you trying to access Spray classes? (in some service/listener/action/in a compiler process/etc...?)

**Reply by moglideveloper** [reply]
No, I am not using any custom class-loaders.

Kindly see next comment to reproduce this issue.

**Comment by mutcianm** [comment]
@moglideveloper is `scala-library` getting packaged in this case like in https://github.com/JetBrains/sbt-idea-plugin/issues/106?

**Reply by moglideveloper** [reply]
yes scala-library is getting packaged.

**Reply by moglideveloper** [reply]
Please find two simplified projects two reproduce this issue.

https://github.com/moglideveloper/two-examples-to-reproduce-linkage-error

Below are the steps to reproduce issue :-

Steps :-   
* Launch plugin in intellij instance
* Create New Project
* Select <b>Sample</b> template and click next
* Now click finish
* check logs on intellij console

**Reply by moglideveloper** [reply]
<img width="793" alt="image" src="https://user-images.githubusercontent.com/8999083/117902615-7d0d1480-b2eb-11eb-9b8a-d786fead6807.png">

<img width="797" alt="image" src="https://user-images.githubusercontent.com/8999083/117902646-8d24f400-b2eb-11eb-819f-6c0ab6372a86.png">

**Comment by unkarjedy** [comment]
@mutcianm 
For some reason in these ^ examples plugin jar is not generated.
`plugin` folder only contains `lib` (with scala-library.jar bundled)
![image](https://user-images.githubusercontent.com/3989292/117937478-b60ab080-b30e-11eb-8e4d-621b1ffb8455.png)

**Comment by mutcianm** [comment]
The classloading constraint violation is caused by the scala-library that gets loaded in both parent(Scala plugin) and child(this plugin) classloaders.
Essentially this should self-fix after I fix #106 which I'm already working on.

**Comment by mutcianm** [comment, marked as answer]
Please try with sbt-idea-plugin version 3.12.4

**Reply by moglideveloper** [reply]
After upgrading to sbt-idea-plugin version 3.12.4 fixes issue.
Many Thanks
